File: elixir/beer-song/beer_song.exs
defmodule BeerSong do
  @doc """
  Get a single verse of the beer song
  """
  @spec verse(integer) :: String.t()
  def verse(number) do
    Enum.reduce(
      [
        &on_the_wall/1,
        &of_beer/1,
        &pass_it_around/1,
        &next_on_the_wall/1
      ],
      "",
      &(&2 <> &1.(number))
    )
  end

  @doc """
  Get the entire beer song for a given range of numbers of bottles.
  """
  def lyrics(), do: lyrics(99..0)
  @spec lyrics(Range.t()) :: String.t()
  def lyrics(range) do
    range
    |> Enum.map(&verse/1)
    |> Enum.join("\n")
  end

  @spec on_the_wall(number) :: String.t()
  def on_the_wall(number) do
    bottles =
      number
      |> get_bottles
      |> String.capitalize()

    "#{bottles} of beer on the wall, "
  end

  @spec of_beer(number) :: String.t()
  def of_beer(number), do: "#{get_bottles(number)} of beer.\n"

  @spec pass_it_around(non_neg_integer) :: String.t()
  defp pass_it_around(0), do: "Go to the store and buy some more"
  defp pass_it_around(1), do: "Take it down and pass it around"
  defp pass_it_around(_), do: "Take one down and pass it around"

  @spec next_on_the_wall(number) :: String.t()
  def next_on_the_wall(number), do: ", #{get_bottles(number - 1)} of beer on the wall.\n"

  @spec pluralize(String.t(), non_neg_integer) :: String.t()
  def pluralize(str, 1), do: "1 #{str}"
  def pluralize(str, count), do: "#{count} #{str}s"

  @spec get_bottles(non_neg_integer) :: String.t()
  defp get_bottles(-1), do: get_bottles(99)
  defp get_bottles(0), do: "no more bottles"
  defp get_bottles(number), do: pluralize("bottle", number)
end

File: clojure/beer-song/src/beer_song.clj
(ns beer-song
  (:require [clojure.string :refer [capitalize join]]))

(defn- bottles [n]
  (let [edge-cases ["no more bottles" "1 bottle"]
        default (str n " bottles")]
    (get edge-cases n default)))

(defn- of-beer [b]
  (str b " of beer.\n"))

(defn- on-the-wall
  ([b]
   (on-the-wall b ", "))
  ([b end]
   (str b " of beer on the wall" end)))

(defn- pass-it [s]
  (str "Take " s " down and pass it around, "))

(defn- take-down [n]
  (let [edge-cases [["Go to the store and buy some more, " 99]
                    [(pass-it "it") 0]]
        default [(pass-it "one") (dec n)]]
    (->> (get edge-cases n default)
         ((fn [[a b]]
            (str
              a
              (on-the-wall (bottles b) ".\n")))))))

(defn verse
  "Returns the nth verse of the song."
  [n]
  (->> n
       ((juxt (comp capitalize on-the-wall bottles)
              (comp of-beer bottles)
              take-down))
       (apply str)))

(defn sing
  "Given a start and an optional end, returns all verses in this interval. If
  end is not given, the whole song from start is sung."
  ([start] (sing start 0))
  ([start end]
   (->> (range end (inc start))
        reverse
        (map verse)
        (join "\n"))))
